README.md
<div align="center">

# AACO_ADB_Auto_Clock_with_OCR
AACO项目开发的原意是使用ADB进行自动打卡，同时加入了OCR识别、超时等待、自动重连和短信通知等特性
<br>自用长期稳定，开源出来贡献给大家，欢迎提交代码！
<br>

<a href="https://twitter.com/Elliot58616851" target="_blank"><img src="https://img.shields.io/twitter/follow/Elliot58616851?style=social"> </a>

<div>
    <img alt="Python" src="https://img.shields.io/badge/python-v3.9-blueviolet">
</div>
<div>
    <img alt="license" src="https://img.shields.io/github/license/elliot-bia/AACO_ADB_Auto_Clock_with_OCR?logo=2">
    <img alt="stars" src="https://img.shields.io/github/stars/elliot-bia/AACO_ADB_Auto_Clock_with_OCR?style=social">
</div>
<br>

[🇨🇳简体中文](./README.md) | [🇺🇸English](./README-en.md) | [🇩🇪Deutsch](./README-de.md)


此项目仅供非营利性学习研究，请勿用作非法用途，如有侵权请及时联系删除

</div>


# 0x00 介绍
上下班打卡是我最讨厌的事情，某天晚上翻来覆去睡不着，想起了[SQLite源码](https://github.com/sqlite/sqlite/blob/master/LICENSE.md)的三句话：
>May you do good and not evil.  
May you find forgiveness for yourself and forgive others.  
May you share freely, never taking more than you give.  

所以我将基于ocr识别的adb打卡程序开源出来，自己用的很稳定，配置简单，兼容性强，适合绝大部分文字识别点击打卡，具有以下特性：
- 中国节假日判断自动打卡
- 基于OCR识别，准确高，反应快
- 加入超时等待（默认4次16+8+4+2秒）
- 加入出错重试（默认3次）
- 出错重试失败后进行短信通知（需要阿里云短信api）
- 加入图像识别范围（纯文字不够用，需要点击图标的情况）
- 配置简单，只有1个配置文件
- 拥有日志、截图辅助判断打卡情况

# 0x01 前期准备
该项目基于安卓的adb打卡，意思就是要有一个手机放公司，或者刷定位模块  
然后开启安卓手机的开发者模式，打开adb，可以用有线连接，也可以无线  
此外该项目还集成了阿里云短信通知，需要可以自行开启  
所需条件：
- 安卓手机（硬件）
- 手机开启adb并正确连接（设置）
- 公司一台开启的电脑（硬件）

前置知识：ADB调试，自行谷歌

# 0x02 打卡原理
我们假设打卡的识别顺序是这样的：
1. 识别文字A(考勤打卡)--确认点击
2. 点击后等待跳转，确认文字B（班打卡）--确认点击成功，流程结束

给个图演示一下  

<div align="left">
<img src="./img/img_2.png" width="25%" height="25%" alt="img_2.png"/><img src="./img/img_3.png" width="25%" height="25%" alt="img_3.png"/>
</div>

给个流程图演示一下  
<img src="./img/img_4.png" width="15%" height="15%" alt="img_4.png"/>

就是这么简单。其中涉及到一些出错重试、手机性能羸弱导致超时等待、多处文字干扰等一些功能，详见最后[其他说明]()  
我们就关注于：__找到需要某个点击的文字（图标）__ --> __确认点击成功的结果__  
我们来举个企业微信打卡例子：
1. 打开企业微信 -->  出现<消息>界面 
2. 点击<工作台> -->  出现<考勤打卡>按钮
3. 点击<考勤打卡>   -->  出现上<班打卡>按钮
4. 点击(打卡图片)   -->  是否<确认>打卡？
5. 点击<确认>打卡按钮   -->  打卡成功

这样就是一个整体的打卡流程, 逻辑是不是很简单？  
__顺序就是打开企业微信-->点击工作台-->点击考勤打卡-->点击打卡（图片）-->点击确定-->打卡成功！__

# 0x03 配置说明
基于KISS原则(Keep It Simple, Stupid),只有一个配置文件需要修改  
将`config.ini.template`改为`config.ini`文件,然后修改配置  
配置文件使用模块化，分为3个部分  
[Config] 默认配置，api、通知等  
[Steps] 每一步的步骤  
[Users] 多用户机器配置

只要添加[Steps]里的配置即可
```editorconfig
[Steps]
# 一共需要4步的点击和判断
TotolSteps = 4

[Step1]
# 选择识别的类型（文字/图片/完成）
Step_Type = TEXT
# 需要点击的文字A
Step_Input = 工作台
# 需要确认点击是否成功的文字B（下同）
Step_Output = 考勤打卡

[Step2]
Step_Type = TEXT
Step_Input = 考勤打卡
Step_Output = 班打卡

[Step3]
# 这里就需要点击图片了
Step_Type = IMAGE
# 这里使用图片的相对路径
Step_Input = /src_img/buoy.png
Step_Output = 确定

[Step4]
# 最后一步识别，不做点击成功判断
Step_Type = FINISH
Step_Input = 确定
```

好，步骤配置到这里就完成大头了，剩下的还要配置一下IP地址、端口、打卡设备名称和短信通知API（可选）

```editorconfig
[Users]
User_list = User1,User2

[User1]
Connect_Type = Wifi
IP_Address = 192.168.99.106
Port = 5555
Device_Name = Rami的小米手机9
# 需要进行短信通知的手机号，留空为不通知
SMS_Phone = 130xxxxxxxx

[User2]
Connect_Type = USB
Series_Strings = xxxxx
Device_Name = Rami的小米手机9
# 需要进行短信通知的手机号，留空为不通知
SMS_Phone = 130xxxxxxxx
```
到这里配置文件基本就修改完成了,如果需要阿里云短信通知，自行放入配置文件里即可

# 0x04 其他配置
## 4.1 可信任设备
在ADB首次连接的时候，会出现是否信任该电脑，允许即可  
## 4.2 ADB不可用
有的手机太旧了，adb版本太新不支持，使用旧版本adb即可


# 0x05 运行
假设你已经完成了上面的
1. adb正确链接手机
2. 配置文件编写完成

那恭喜你，可以开始尝试运行脚本了！
```shell
# 以ubuntu为例
apt install android-tools-adb python3-pip build-essential libgl1-mesa-glx

# 克隆仓库
git clone --depth=1 https://github.com/elliot-bia/AACO_ADB_Auto_Clock_with_OCR.git
cd AACO_ADB_Auto_Clock_with_OCR
# 安装pytorch, 可以参考https://pytorch.org/进行更改
pip3 install torch torchvision torchaudio --index-url https://download.pytorch.org/whl/cpu
pip install cnocr -i https://pypi.doubanio.com/simple 
pip install opencv-python pure-python-adb alibabacloud_dysmsapi20170525==2.0.23 chinesecalendar aircv  -i https://pypi.tuna.tsinghua.edu.cn/simple  

# 运行
python ./bin/main.py
```
如果感觉无从下手，可以请我帮你远程指导，只需要打赏¥200即可购买我的一小时的帮助（牛郎在世🤣  
或者参考[Wiki](https://github.com/elliot-bia/AACO_ADB_Auto_Clock_with_OCR/wiki)

# 0x06 特别鸣谢
图像识别[aircv](https://github.com/NetEaseGame/aircv)  
文字识别[cnocr](https://github.com/breezedeus/CnOCR)   
工作日判断[chinese_calendar](https://github.com/LKI/chinese-calendar)    
adb操作[pure-python-adb](https://pypi.org/project/pure-python-adb/)  
numpy   
opencv-python   

# 0x07 捐赠
赏点钱喝杯咖啡吧~  
<img src="img/zfb_wx_qc.png" width="60%" height="60%" alt="zfb_wx_qc.png"/>


Changelog.md
# 待更新功能/Todo
- Using minitouch replace ppadb, added touch pressure function.
- 用minitouch替代ppadb，增加触摸压力功能，绕过检测，更加模拟真人
> Authro: Rami


# Version V5 2023-05-10 19:54:36
## En
- Change config file, make User_list as list type
- Change log configuration, make it easier to check
- Adding virtual environment configuration
## Cn
- 修改config配置文件用户列表生效为list类型
- 修改log配置
- 增加虚拟环境配置

# 版本更新V4 2023-03-05 02:05:42
将代码大幅度优化和重构，重新编写Wiki，将代码逻辑重新理顺，并将其开源

# 版本更新V3  2022-11-22 16:46:01
新增功能：
- [x] adb连接失效的时候自动提醒
- [x] 增加短信提醒功能
- [x] 使用ocr识别坐标，然后点击
- [x] 写出一个集合或者类，使用ocr坐标即可点击
- [x] 加入错误尝试功能，然后提醒
- [x] 优化逻辑，~~尝试使用装饰器~~，免得日志太乱
- [x] 日志要分开，日志要分级
- [x] 截图保存错误日志
- [x] 新增多连接功能



test/test.py

from pathlib import Path
import sys
BASE_DIR = str(Path(__file__).resolve().parent.parent)
sys.path.append(BASE_DIR)

from configparser import ConfigParser
import sys

config = ConfigParser()
config['DEFAULT'] = {'Pool': '4'}

with open(sys.path[-1] + '/conf/config.ini', 'w',  encoding='utf-8') as configfile:
    config.write(configfile)

test/ConfigGenerate.py

from configparser import ConfigParser
import sys

config = ConfigParser()
config['DEFAULT'] = {'Key': 'Value'}

print(sys.path[-1])
with open(sys.path[-1] + '/conf/config.ini', 'w',  encoding='utf-8') as configfile:
    config.write(configfile)

module/sms_notify/__init__.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# date：  2022-11-23 09:59:14
# author  Rami
#
"""
本文件描述:
阿里云短信通知接口
注意短信使用次数

用法：
from module import sms_notify

sms_notify.sms_monitor_send(phone_numbers='130xxxx',template_param='{"content":"%s"}' %("new_content"))

注意：
会自动在base dir目录，也就是sys.path[-1]下，创建log目录
"""

from alibabacloud_dysmsapi20170525.client import Client as Dysmsapi20170525Client
from alibabacloud_tea_openapi import models as open_api_models
from alibabacloud_dysmsapi20170525 import models as dysmsapi_20170525_models
from alibabacloud_tea_util import models as util_models
from alibabacloud_tea_util.client import Client as UtilClient

from pathlib import Path
import sys
from module import log_factory
BASE_DIR = str(Path(__file__).resolve().parent.parent.parent)
sys.path.append(BASE_DIR)
import configparser

def create_client(access_key_id: str,access_key_secret: str,) -> Dysmsapi20170525Client:
    """
        使用AK&SK初始化账号Client
        @param access_key_id:
        @param access_key_secret:
        @return: Client
        @throws Exception
        """
    config = open_api_models.Config(
        # 您的AccessKey ID,
        access_key_id=access_key_id,
        # 您的AccessKey Secret,
        access_key_secret=access_key_secret
    )
    # 访问的域名
    config.endpoint = f'dysmsapi.aliyuncs.com'
    return Dysmsapi20170525Client(config)

def sms_monitor_send(phone_numbers,template_param='{"code":"1234"}') -> None:
    config = configparser.ConfigParser()
    config.read(BASE_DIR + '/conf/config.ini')
    access_key_id = config['AliyunSMS']['Access_Key_ID']
    access_key_secret = config['AliyunSMS']['access_key_secret']
    sign_name = config['AliyunSMS']['sign_name']
    template_code = config['AliyunSMS']['template_code']

    if access_key_id == "":
        # 如果不填，直接返回
        return

    try:
        client = create_client(access_key_id, access_key_secret)
        send_sms_request = dysmsapi_20170525_models.SendSmsRequest(
            sign_name=sign_name,
            template_code=template_code,
            phone_numbers=phone_numbers,
            template_param=template_param
        )
    except Exception as e:
        log_factory.DEBUGGER.error(e,exc_info=True)

    runtime = util_models.RuntimeOptions()
    try:
        # 复制代码运行请自行打印 API 的返回值
        client.send_sms_with_options(send_sms_request, runtime)
        log_factory.DEBUGGER.info(f"短信发送成功！短信号码：{phone_numbers}，短信内容{template_param}")
    except Exception as error:
        # 如有需要，请打印 error
        log_factory.DEBUGGER.debug(error, exc_info=True)  
        messages = UtilClient.assert_as_string(error.message)
        log_factory.DEBUGGER.error(f"\n\n{'='*30}\n\n短信发送出错，官方出错代码为：{messages}\n\n{'='*30}\n\n", exc_info=True)  

if __name__ == '__main__':



    if not sys.version_info >= (3, 0):
        sys, exit('[x] WARNING - this script requires Python 3.x  Exiting')
    
    # 测试代码
    sms_monitor_send(phone_numbers='123456778',template_param='{"content":"%s"}' %("new_content"))

core/PhoneCtrl.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# date：  2022-11-23 09:59:14
# author  Rami
#
"""
本文件描述:

定义 PhoneCtrl类，执行功能有：
PhoneCtrl.conn()
PhoneCtrl.clock_in
PhoneCtrl.clock_off

加入retry、跟短信告警

"""

from ppadb.client import Client as AdbClient
import time
from pathlib import Path
import sys
BASE_DIR = str(Path(__file__).resolve().parent.parent)
sys.path.append(BASE_DIR)
# 导入本地包
from module import log_factory
# 导入图像识别包
import aircv as ac
# 导入crc文字识别包
from cnocr import CnOcr
import numpy as np
import configparser


class PhoneAdbCtrl():
    def __init__(self,User) -> None:
        self.config = configparser.ConfigParser()
        self.config.read(BASE_DIR + '/conf/config.ini')
        self.name = self.config[User]['Device_Name']
        self.client = AdbClient(host="127.0.0.1", port=5037)
        if self.config[User]['Connect_Type'] == 'Wifi':
            self.ip = self.config[User]['IP_Address']
            self.port = int(self.config[User]['Port'])
            self.phone = self.client.device(str(self.ip) + ':' + str(self.port))
        elif self.config[User]['Connect_Type'] == 'USB':
            Series_Strings = self.config[User]['Series_Strings']
            self.phone = self.client.device(Series_Strings)

        self.OCR_Timeout_Retry = self.config.getint('Config','OCR_Timeout_Retry')




    def send_tap(self,x,y):
        self.phone.input_tap(x,y)

    def img_match(self,imsrc_path,imsch_path):
        imsrc = ac.imread(imsrc_path) # 原始图像
        imsch = ac.imread(imsch_path) # 带查找的部分
        ac_postion = ac.find_template(imsrc, imsch)

        log_factory.DEBUGGER.debug(f"查找源{imsrc_path}，目标{imsch_path}，成功匹配到图像！位置：{ac_postion}")
        return ac_postion


    def text_identify(self,text,img_path):
        """
        传入需要识别的文字和图片，返回点击坐标
        
        如果返回为-1, 则是未识别图片

        TODO:尽量不要出现多个同样的text
        """
        
        img_fp = img_path
        ocr = CnOcr() 
        out = ocr.ocr(img_fp)
        axis_postion = "-1"
        for i in out:
            if text in i['text']:
                axis_postion = np.mean(i['position'],axis=0)
                # print(axis_postion)
                log_factory.DEBUGGER.debug(f"成功获取坐标！位置：{axis_postion}")

        return axis_postion


    def screenshot(self,level="info",save_path="/screenshot/"):
        """
        进行截图，如果是error等级的，会进行重命名
        如果是info等级，处理结束后会使用正则进行匹配命名删除

        一个机器一次只处理一张截图，返回截图路径+文件名
        """
        if level == "error":
            save_path = sys.path[-1] + save_path + self.name + "-" + self.photo_name + "-error.png"
        else:
            save_path = sys.path[-1] + save_path + self.name + "-" + self.photo_name + "-info.png"
        log_factory.DEBUGGER.debug(f"进行设备{self.name}的{self.photo_name}过程截图")
        result = self.phone.screencap()
        with open(save_path, "wb") as fp:
            fp.write(result)
        return save_path

    def send_keyevent(self, keycode):
        """
        发送一个按键事件
        https://developer.android.com/reference/android/view/KeyEvent.html
        :return:
        """
        self.phone.input_keyevent(keycode)

    def try_click(self):
        """
        尝试点击确认、确定、允许、继续 这几个字眼, 加上稍后
        """
        result = self.text_identify("稍后",self.screenshot_path)
        if result != "-1":
            log_factory.DEBUGGER.info(f"设备{self.name}的{self.photo_name}过程截图尝试点击《稍后》")
            self.send_tap(result[0],result[1])
        result = self.text_identify("确认",self.screenshot_path)
        if result != "-1":
            log_factory.DEBUGGER.info(f"设备{self.name}的{self.photo_name}过程截图尝试点击《确认》")
            self.send_tap(result[0],result[1])
        result = self.text_identify("确定",self.screenshot_path)
        if result != "-1":
            log_factory.DEBUGGER.info(f"设备{self.name}的{self.photo_name}过程截图尝试点击《确定》")
            self.send_tap(result[0],result[1])
        result = self.text_identify("允许",self.screenshot_path)
        if result != "-1":
            log_factory.DEBUGGER.info(f"设备{self.name}的{self.photo_name}过程截图尝试点击《允许》")
            self.send_tap(result[0],result[1])
        result = self.text_identify("继续",self.screenshot_path)
        if result != "-1":
            log_factory.DEBUGGER.info(f"设备{self.name}的{self.photo_name}过程截图尝试点击《继续》")
            self.send_tap(result[0],result[1])
        pass




    def check_right(self,text):
        """
        此函数用做超时检查，避免因为机型性能问题导致等待时间过长

        传入需要识别的文字/图片，进行判断，超时2的指数重试5次，重试3次后就重头开始

        同时如果出现 确认、确定、允许三个字眼的时候，进行尝试点击
        """

        time_sleep = 1
        for i in range(0, self.OCR_Timeout_Retry):
            # i为等待超时
            self.screenshot_path = self.screenshot()
            result = self.text_identify(text,self.screenshot_path)
            if result != "-1":
                log_factory.DEBUGGER.info(f"进行设备{self.name}的{self.photo_name}过程截图识别成功,坐标{result}")
                return
            elif result == "-1":
                log_factory.DEBUGGER.warning(f"进行设备{self.name}的{self.photo_name}过程截图识别不成功,进行重试第{i}次中！")
                log_factory.DEBUGGER.info(f"等待{time_sleep}秒中")
                self.try_click()
                time.sleep(time_sleep)
                time_sleep *= 2
            else:
                pass
                log_factory.DEBUGGER.error(f"{self.name}设备无法打开{self.photo_name},重新尝试")

        self.screenshot(level="error")
        log_factory.DEBUGGER.error(f"{self.name}设备无法打开{self.photo_name},超时等待无效，进行重试")
        return "-1"


            
    def final_tap_check(self, Input, Step_Section):
        text = Input
        self.photo_name= Step_Section
        log_factory.DEBUGGER.debug(f"设备{self.name}进行点击最后步骤{Step_Section},点击目标为{Input}")

        # 确定需要等待咋办
        self.check_right(text)
        # 有两个确定咋办？
        self.screenshot_path = self.screenshot()

        img_fp = self.screenshot_path
        ocr = CnOcr() 
        out = ocr.ocr(img_fp)
        axis_postion = []
        for i in out:
            # 返回存在的坐标
            if text in i['text']:
                axis_postion.append(np.mean(i['position'],axis=0))
                # print(axis_postion)
                log_factory.DEBUGGER.debug(f"成功{text}获取坐标！第{i}个位置：{axis_postion}")
        for j in axis_postion:
                self.phone.input_tap(j[0],j[1])
                log_factory.DEBUGGER.info(f"{self.name}设备在截图{self.photo_name}中尝试点击{text}")
                result = self.text_identify("正常",self.screenshot_path)
                if result == "-1":
                    log_factory.DEBUGGER.warning(f"{self.name}设备在截图{self.photo_name}中点击{text}未成功，继续")
                    continue
                else:
                    pass
        
        self.photo_name="FINASH"
        self.screenshot()
        log_factory.DEBUGGER.info(f"{self.name}设备打卡成功！")


    def tap_button(self):
        self.photo_name="tap_button"
        log_factory.DEBUGGER.debug(f"设备{self.name}尝试点击打开，进行图片匹配")
        for i in range(0,3):
            return_postion = self.click_and_check_img(target= sys.path[-1] + "/src_img/buoy.png",click_type="img")
            if return_postion == None:
                log_factory.DEBUGGER.warning(f"设备{self.name}进行图片匹配未发现，尝试下滑{i}次")
                self.phone.input_swipe(540, 1300, 540, 500, 100)
                continue
            else:
                x = return_postion['result'][0]
                y = return_postion['result'][1]
                log_factory.DEBUGGER.debug(f"设备{self.name}进行图片匹配成功，坐标{x}，{y}")
                self.send_tap(x,y)
                break

            





    def retrun_img_match_position(self,target):
        self.screenshot_path = self.screenshot()
        imsrc = ac.imread(self.screenshot_path) # 原始图像
        imsch = ac.imread(target) # 带查找的部分
        click_axios = ac.find_template(imsrc, imsch)
        log_factory.DEBUGGER.debug(f"查找源{self.screenshot_path}，目标{target}，成功匹配到图像！位置：{click_axios}")
        return click_axios


    def click_and_check_txt(self,Input,Output,Step_Section):
        """
        先对当前页面进行截图
        然后判断Input，点击，如果无法成功(超时OCR_Timeout_Retry次数)，就认为识别失败，返回TimeoutError
        往下滑动Downslide_Retry次，重复识别check_right，识别错误则抛出Exception
        检查Ouput是否点击成功
        """
        self.photo_name=Step_Section
        log_factory.DEBUGGER.info(f"设备{self.name}开始打卡步骤{Step_Section},类型为『TXT』，打卡关键字为{Input}，打卡成功关键字为{Output}")
        for i in range(1,self.config.getint('Config','Downslide_Retry')+2): # range是用1，2，3不使用
            # 进行截图并返回路径
            self.screenshot_path = self.screenshot()
            click_axios = self.text_identify(Input, self.screenshot_path)
            self.send_tap(click_axios[0], click_axios[1])
            if self.check_right(Output) != "-1":
                log_factory.DEBUGGER.info(f"{self.name}设备尝试打开{Output}成功，成功找到{Output}!")
                # 成功找到就跳出循环
                break
            elif self.check_right(Output) == "-1":
                log_factory.DEBUGGER.warning(f"{self.name}设备打卡步骤{Step_Section}识别Input{Input}和Output{Output}失败，进行下滑，次数{i}次!")
                self.phone.input_swipe(540, 1300, 540, 500, 100)
                # 没有找到就继续咯
            if i == self.config.getint('Config','Downslide_Retry')+1:
                # 尝试最后都没匹配到，抛出识别错误
                raise Exception(f"{self.name}设备打卡步骤{Step_Section}识别Input{Input}和Output{Output}失败，进行下滑次数{i}次均失败!,抛出错误")



    def click_and_check_img(self,IMG_Input, Output, Step_Section):
        self.photo_name=Step_Section
        log_factory.DEBUGGER.info(f"设备{self.name}开始打卡步骤{Step_Section},类型为『IMAGE』，打卡关键图像为{IMG_Input}，打卡成功关键字为{Output}")
        for i in range(1,self.config.getint('Config','Downslide_Retry')+2):
            return_postion = self.retrun_img_match_position(target= sys.path[-1] + IMG_Input)
            time.sleep(2)
            if return_postion == None:
                log_factory.DEBUGGER.warning(f"设备{self.name}进行图片匹配未发现，尝试下滑{i}次")
                self.phone.input_swipe(540, 1300, 540, 500, 100)
                
            else:
                x = return_postion['result'][0]
                y = return_postion['result'][1]
                log_factory.DEBUGGER.info(f"设备{self.name}进行图片匹配成功，坐标{x}，{y}")
                self.send_tap(x,y)
                break
            if i == self.config.getint('Config','Downslide_Retry')+1:
                # 尝试最后都没匹配到，抛出识别错误
                raise Exception(f"{self.name}设备打卡步骤{Step_Section}识别Input{IMG_Input}和Output{Output}失败，进行下滑次数{i}次均失败!")


    def open_wecom(self):
        # self.phone.shell("am start com.tencent.mm/com.tencent.mm.ui.LauncherUI")
        self.phone.shell("am force-stop com.tencent.wework")
        time.sleep(0.5)
        self.phone.shell("am start com.tencent.wework/com.tencent.wework.launch.LaunchSplashActivity")
        time.sleep(0.5)

        self.photo_name="open_wecom"

        log_factory.DEBUGGER.debug(f"设备{self.name}尝试打开微信，判断关键字：消息")
        self.check_right("消息")



    def open_phone(self):
        if "false" in self.phone.shell("dumpsys deviceidle | grep mScreenOn"):
            log_factory.DEBUGGER.debug(f"设备{self.name}未亮屏，进行亮屏")
            self.send_keyevent("KEYCODE_POWER")
        self.send_keyevent("KEYCODE_HOME")
        time.sleep(1)
        self.photo_name="open_phone"
        self.screenshot()
        pass


if __name__ == '__main__':


    if not sys.version_info >= (3, 0):
        sys, exit('[x] WARNING - this script requires Python 3.x  Exiting')
    
    # 测试代码
    # log_factory.DEBUGGER.debug(f"设备未亮屏，进行亮屏")

